### PR TITLE
first pass of optimizer

### PR DESCRIPTION
**Pull request by tobymao**
initial optimizer rule, to rewrite sql into canonical form

useful for parsing arbitrary queries and physical plans

### CODE REVIEW

File: sqlglot/helper.py [hunk 1]
```diff
@@ -62,6 +62,15 @@ def apply_index_offset(expressions, offset):
     return expressions
 
 
+def parent(expression, kind):
```
Comment (nolamesa) [inline sqlglot/helper.py:65]: This is not currently used anywhere right? do you need it? If so perhaps rename to find_ancestor. Also minor but I would change the local variable below from expression to ancestor and return None when not found

File: sqlglot/optimizer.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        return node
+
+    return expression.transform(qualify)
+
+
+def qualify_columns(expression, _schema):
+    return expression
+
+
+def optimize(expression, schema=None, db=None, catalog=None):
```
Comment (nolamesa) [inline sqlglot/optimizer.py:21]: I am not entirely clear how qualifying tables is an optimization

**Review by nolamesa — APPROVED** [review]

File: sqlglot/optimizer.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/optimizer.py:21]: I think this is a _pre_ optimization step.

File: sqlglot/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            if not node.args.get("catalog"):
+                node.set("catalog", catalog)
+            return node
+        return node
+
+    return expression.transform(qualify)
+
+
+def qualify_columns(expression, _schema):
+    return expression
```
Comment (barakalon) [inline sqlglot/optimizer.py:18]: What were you imaging here?
I was thinking DFS in postorder, disambiguating columns as we go up the tree?

**Review by barakalon — APPROVED** [review]
Sweet!

I think this canonical SQL will be helpful for the Minerva SQL proxy. But that proxy doesn't have to actually generate/optimize a logical plan.

For a full optimizer, I think we'll need to convert an expression to a logical plan?

File: sqlglot/optimizer.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/optimizer.py:21]: right, there are a series of steps needed to take every query into canonical form, this is the first rule

**Review by nolamesa — APPROVED** [review]

**Pull request merged**